### DISCUSSION TITLE
Migrating from Valetudo 0.6 to Valetudo RE

### DISCUSSION THREAD

**Discussion by xefil** [rand256/valetudo#379] [General]
Hi all,
I'm using Valetudo 0.6 on my Gen1 Xiaomi Vacuum (rockrobo.vacuum.v1) with firmware 'v11_004018.12edfghb75.fullos.pkg' and would like to test Valetudo RE.
So I have some questions:

1. First of all, which procedure should I follow to apply your mod?

2. Is then possible to implement forbidden zones or virtual walls in any way (or workarounds?) ?

3. I've then an issue I hope this version could solve. The Vacuum is under the sofa. Sometims happens that the sofa is so low that the Vacuum draws it and in case of Zone Cleaning it doesn't go out from the sofa due it's thinking the sofa is a wall. I need to manually control it out from the sofa (i.e. go forward for 50cm) and THEN start a zone cleaning. In a similar situation if I clean a zone which has the door closed (because I've rabbits around the home :-D ), if I want that later the robot goes to the other room, due it has recorded there was a closed door, he cannot find the correct way to reach it. Only solution, in this case, is to control it manually again through the open door. The way to home works always.

Any suggestions and hints?

Thanks a lot!

Simon

**Comment by rand256** [comment]
1. Just for trying on Gen1 you may want to download just valetudo.tar.gz packed binary to `/mnt/data/valetudo`, then unpack it, run `stop valetudo` to stop currently running version and finally run RE with `./valetudo` from that directory. After you've finished with your testing, simply press `ctrl+C` to stop RE and type `start valetudo` to run again your installed version. Oh, and keep a copy of your existing configuration - `/mnt/data/valetudo/config.json` file - in case something breaks.
For a permanent installation I'd suggest flashing a pre-built image with pre-installed RE binary (or build it yourself using _vacuumz_ image builder).

2. The thing is that valetudo is actually something like a GUI (Web UI) working over a low-level manufacturer's firmware. So it can only call functions the device already have. Plus a few hacks like copying maps (which is possible only because maps are actually files which can be copied). To create forbidden zones or virtual walls you need a firmware from gen2 running on gen1 - and it's not about valetudo.

3. For this you can save a map with all doors opened and reload it when you need it.

**Reply by xefil** [reply]
Thanks for detailed answer @rand256 . I think I'll give it a try. The backup-restore map function is something I really need. Every time my girfriend press the button on the vaccum and voilà, the map is lost :'(
A last question (for now). Is there a way to move the robot in a position like we do using the manual control, from Home Assistant? I mean: in the sofa example, if I move the robot out from the sofa, it's possible to let him move around the house, otherwise it happens he thinks the sofa is a wall (As descrived above). If I could automate a movement straight forward for one meter, without it checks whether there are walls or not, then in sequence I could let him run to a zone clean. Ideas? Otherwise I would need to run it a first time without sofa and, like descrived by you, reload the map every time.

Here's my sofa and a test with GoTo option. You can see how the robot tries to go out from the sofa not in a straight way.

![image](https://user-images.githubusercontent.com/17131053/106522353-e3f68280-64df-11eb-8f31-8c57af02a23b.png)


Thanks!!
Simon

**Reply by rand256** [reply]
Actually you can try to emulate in HA what's going on at manual control tab. See [this wiki page](https://github.com/rand256/valetudo/wiki/REST-API), you'll need URLs `/api/start_manual_control`, `/api/stop_manual_control` and `/api/set_manual_control`. The angle and velocity numeric range for the latter you can check on the control tab, and the sequenceId is just a random sequential numbers.

**Reply by xefil** [reply]
Thanks a lot! I'm testing it based on your hints and I'll install it asap. The auto-map-restore option sounds great and is something that could help me a lot.
About virtual walls, well, I'll try to create more zones excluding possible frequent forbidden areas.
And for the manual mode, sounds good. @rand256 seems that this version could help me a lot!
Simon

PS: what does exactly save the map-backup?

**Reply by rand256** [reply]
> what does exactly save the map-backup?

For gen1 it just copies [these](https://github.com/rand256/valetudo/blob/78ece70d4c20b61821e5a043e941df1e702f5752/lib/MapManager.js#L5-L6) files and restarts some system processes. In gen2+ there are a few less barbaric ways to reload the map, but they all provide the same results.

**Reply by xefil** [reply]
Hello @rand256
ok for the backup, thank you.
I've tried to update via miio as done with previous version like this:

`mirobo --token MyToken--ip 192.168.3.61 update-firmware /root/software/vacuum/Valetudo_RE/v11_004018.pkg`

But at the end of the download, at 100% il fails. Ideas?

```
Downloading:  99%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋ | 99/100 [01:47<00:01,  1.28s/it]DEBUG:miio.miioprotocol:192.168.3.61:54321 >>: {'id': 1444, 'method': 'miIO.get_ota_state', 'params': []}
DEBUG:miio.miioprotocol:192.168.3.61:54321 (ts: 2021-02-02 11:25:02, id: 1444) << {'result': ['downloading'], 'id': 1444}
DEBUG:miio.miioprotocol:192.168.3.61:54321 >>: {'id': 1445, 'method': 'miIO.get_ota_progress', 'params': []}
DEBUG:miio.miioprotocol:192.168.3.61:54321 (ts: 2021-02-02 11:25:02, id: 1445) << {'result': [99], 'id': 1445}
Downloading:  99%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋ | 99/100 [01:48<00:01,  1.28s/it]DEBUG:miio.miioprotocol:192.168.3.61:54321 >>: {'id': 1446, 'method': 'miIO.get_ota_state', 'params': []}
DEBUG:miio.miioprotocol:192.168.3.61:54321 (ts: 2021-02-02 11:25:03, id: 1446) << {'result': ['failed'], 'id': 1446}
DEBUG:miio.miioprotocol:192.168.3.61:54321 >>: {'id': 1447, 'method': 'miIO.get_ota_progress', 'params': []}
DEBUG:miio.miioprotocol:192.168.3.61:54321 (ts: 2021-02-02 11:25:03, id: 1447) << {'result': [99], 'id': 1447}
Failed:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌ | 99/100 [01:50<00:01,  1.12s/it]
DEBUG:miio.vacuum_cli:Writing {'seq': 1447, 'manual_seq': 0} to /root/.cache/python-miio/python-mirobo.seq
(venv) root@cctv:~/software/vacuum/flasher#
```

Thanks

**Reply by rand256** [reply]
If you have 4018 fw installed without special patches then you can't flash firmware this way anymore - the manufacturer forbids that in recent versions. In this case you need to do a factory reset to downgrade, and then try again.

**Reply by xefil** [reply]
Ok, I'll give it a try with a factory reset and then reflash it. I'll leto you know.
Thanks
